tracing/tracing.go

```diff
@@ -0,0 +1,124 @@
+	"google.golang.org/grpc"
+)
+
+// InitOpenTelemetry read config and initializes otel middleware, sets the exporter
+// propagator and global tracer provider.
+func InitOpenTelemetry(config *srvconfig.Config) (func(), error) {
+	ctx := context.Background()
+
+	// Check if tracing is configured
+	if config.OpenTelemetry == *new(srvconfig.OpenTelemetryConfig) {
+		logrus.Info("OpenTelemetry configuration is not found in config, tracing will be disabled")
+		return nil, nil
+	}
+
+	// Validate configuration
+	err := config.OpenTelemetry.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid open telemetry configuration")
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			// Service name used to displace traces in backends
+			semconv.ServiceNameKey.String("containerd"),
+		),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create resource")
+	}
+
+	// Configure OTLP trace exporter and set it up to connect to OpenTelemetry collector
+	// running on a local host.
+	ctrdTraceExporter, err := otlptracegrpc.New(ctx,
+		otlptracegrpc.WithEndpoint(config.OpenTelemetry.ExporterEndPoint),
+		otlptracegrpc.WithDialOption(grpc.WithBlock()),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create trace exporter")
+	}
+
+	// Register the trace exporter with a TracerProvider, using a batch span
+	// process to aggregate spans before export.
+	ctrdBatchSpanProcessor := sdktrace.NewBatchSpanProcessor(ctrdTraceExporter)
+	ctrdTracerProvider := sdktrace.NewTracerProvider(
+		// We use TraceIDRatioBased sampling. Ration read from config translated into following
+		// if sampleing ration < 0 it is interpreted as 0. If ration >= 1, it will always sample.
+		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(config.OpenTelemetry.TraceSamplingRatio)),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(ctrdBatchSpanProcessor),
+	)
+	otel.SetTracerProvider(ctrdTracerProvider)
+
+	// set global propagator to tracecontext
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+
+	return func() {
+		// Shutdown will flush any remaining spans and shut down the exporter
+		err := ctrdTracerProvider.Shutdown(ctx)
+		if err != nil {
+			logrus.Errorf("failed to shutdown TracerProvider: %v", err)
+		}
+
+	}, nil
+}
+
+// StartSpan starts child span in a context.
+func StartSpan(ctx context.Context, opName string, opts ...trace.SpanStartOption) (trace.Span, context.Context) {
+	parentSpan := trace.SpanFromContext(ctx)
+	tracer := trace.NewNoopTracerProvider().Tracer("")
+	if parentSpan.SpanContext().IsValid() {
+		tracer = parentSpan.TracerProvider().Tracer("")
+	}
+	ctx, span := tracer.Start(ctx, opName, opts...)
+	return span, ctx
+}
+
+// StopSpan ends the span specified
+func StopSpan(span trace.Span) {
+	span.End()
+}
+
+// GetContextWithSpan sets the span of a context from another context if not set already
+func GetContextWithSpan(ctx, ctxAnother context.Context) context.Context {
```

Comment:
- GetXxx shouldn't *set* things :)
- Seems this one isn't used now.

Reply:
Changed wording. And yes it's not used immediately but next commit will. Let me know if removing is better idea.

Reply:
Since it's not used, can you remove it from this PR?